types/src/check_summed_hex.rs

```diff
@@ -0,0 +1,436 @@
+//! Check-summed hex encoding following an [EIP-55][1]-like scheme.
+//!
+//! Contains a [serde] helper trait [CheckSummedHex] which is based on [`hex_buffer_serde::Hex`][2].
+//!
+//! [1]: https://eips.ethereum.org/EIPS/eip-55
+//! [2]: https://docs.rs/hex-buffer-serde/0.3.0/hex_buffer_serde/trait.Hex.html
+use alloc::{borrow::Cow, string::String, vec::Vec};
+use base16;
+use core::{convert::TryFrom, fmt, marker::PhantomData};
+
+use blake2::{Blake2b, Digest};
+use serde::{
+    de::{Error as DeError, Unexpected, Visitor},
+    Deserializer, Serializer,
+};
+
+/// Takes a slice of bytes and breaks it up into a vector of *nibbles* (ie, 4-bit values)
+/// represented as `u8`s.
+fn bytes_to_nibbles(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item = u8> + '_ {
+    input
+        .as_ref()
+        .iter()
+        .flat_map(move |byte| [4, 0].iter().map(move |offset| (byte >> offset) & 0x0f))
+}
+
+/// Takes a slice of bytes and outputs an infinite cyclic stream of bits for those bytes.
+fn bytes_to_bits_cycle(bytes: Vec<u8>) -> impl Iterator<Item = bool> {
+    bytes
+        .into_iter()
+        .cycle()
+        .flat_map(move |byte| (0..8usize).map(move |offset| ((byte >> offset) & 0x01) == 0x01))
+}
+
+/// Computes a Blake2b hash outputs it as a vector of *nibbles* (ie, 4-bit values) represented as
+/// `usize`s.
+fn blake2b_hash(data: impl AsRef<[u8]>) -> Vec<u8> {
+    let mut hasher = Blake2b::new();
+    hasher.update(data);
+    hasher.finalize().to_vec()
+}
+
+const HEX_CHARS: [char; 16] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
+];
+
+/// Encodes bytes as hexadecimal with mixed-case based checksums following a scheme similar to
+/// [EIP-55][1].
+///
+/// Key differences:
+///   - Works on any length of data, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Uses hash bits rather than nibbles
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn encode(input: &(impl AsRef<[u8]> + ?Sized)) -> String {
```

Comment:
Maybe we should have this be lazy, like

```suggestion
pub fn encode(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item=char> {
```

Reply:
We might want two of these: one lazy and one not.  Or maybe we can use some hints with sizes for the iterator...